web/github.go

```diff
@@ -0,0 +1,109 @@
+package web
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/bradleyfalzon/ghinstallation"
+	"github.com/google/go-github/github"
+	"github.com/pressly/lg"
+)
+
+// GitHub is an HTTP service to call GitHub endpoints
+type GitHub struct {
+	AppID      int
+	PrivateKey string
+}
+
+func (g *GitHub) installations() ([]*github.Installation, error) {
+	appTr, err := ghinstallation.NewAppsTransportKeyFromFile(
+		http.DefaultTransport, g.AppID, g.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	appClient := github.NewClient(&http.Client{Transport: appTr})
+
+	var installations []*github.Installation
+	opts := &github.ListOptions{
+		PerPage: 100,
+	}
+	for {
+		installs, resp, err := appClient.Apps.ListInstallations(context.TODO(), opts)
+		if err != nil {
+			return nil, err
+		}
+
+		installations = append(installations, installs...)
+
+		if resp.NextPage == 0 {
+			break
+		}
+
+		opts.Page = resp.NextPage
+	}
+
+	return installations, nil
+}
+
+// orgResponse is the response type used by Orgs handler
+type orgRespose struct {
+	Name string `json:"name"`
+}
+
+// Orgs writes in the response the list of organizations where the
+// logged-in user is an admin
+func (g *GitHub) Orgs(w http.ResponseWriter, r *http.Request) {
+	login, err := GetUserLogin(r.Context())
+	if err != nil {
+		lg.RequestLog(r).Errorf("failed to get user login from context: %s", err)
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+
+	installations, err := g.installations()
+	if err != nil {
+		lg.RequestLog(r).Errorf("failed to retrieve the GitHub App installations: %s", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	// initialized as empty array because otherwise json response will be null
+	// instead of []
+	orgs := []orgRespose{}
+
+	for _, installation := range installations {
+		// New transport for each installation
+		itr, err := ghinstallation.NewKeyFromFile(
+			http.DefaultTransport, g.AppID, int(installation.GetID()), g.PrivateKey)
+
+		if err != nil {
+			lg.RequestLog(r).Errorf("failed to initialize the GitHub App installation client: %s", err)
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// Use installation transport in a new client
+		client := github.NewClient(&http.Client{Transport: itr})
+
+		if installation.Account == nil || installation.Account.Login == nil {
+			lg.RequestLog(r).Errorf("failed to get GitHub installation organization name")
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		org := *installation.Account.Login
+		mem, _, err := client.Organizations.GetOrgMembership(context.TODO(), login, org)
```

Comment:
do we actually have permissions to do it? I don't think we request this permission for our app:
<img width="758" alt="screenshot 2019-01-11 at 13 51 56" src="https://user-images.githubusercontent.com/406916/51034788-2589fd00-15a8-11e9-89d3-09292dcdf1af.png">

I also not sure if we really should request it.

Reply:
Ah, I changed the permissions in my test app and forgot to add it to the docs in this PR.

I think we can add a note in https://github.com/src-d/lookout/blob/master/docs/web.md, and open an issue to add it in https://github.com/src-d/lookout/blob/master/docs/configuration.md#authentication-as-a-github-app when we really include it in a release.

What do you mean by this?
> I also not sure if we really should request it.

Do we have another alternative to see if the user in an admin in the organization?

Reply:
The alternative is to ask `read:org` scope when a user authorizes in web UI. 

I'm not sure which way is better (not from the code point of view but from user one). Most probably if web UI is optional then this extra permission request better be optional too when you actually use web UI.

I'm good if for now we keep asking this permission in the App. But please consider the alternative.

Reply:
Off the top of my head, what you propose would be done once, when you log-in, right? And then we would need to store the list of orgs somewhere. Since we said we didn't want to have user management on the DB for now, it would go in the JWT I guess.

But then you have a list that is not up to date, and we would have to compromise:
- assume you need ask the user to manually log out and log in to see new organizations
- have a low expiration date in the JWT, which could be annoying for users asking to log in too often

It's not so clear to me how well it would work, and it does not look a trivial change. My proposal is to continue with the mechanism in this PR, as it was defined in the design doc, and discuss your proposal later. If we see it has advantages it can be switched later.
For now I added the new permission requirement in fe6cfa4.

Reply:
we have a valid token oauth in JWT already. We use it in `/me` endpoint that is called everytime you open UI. You can call orgs endpoint with the same endpoint. No need to cache and it will expired at the same time as `/me`.

Reply:
:thinking: somehow I missed that and I was convinced we were only storing the user id; now it's an easy change. Still, I would prefer to have this PR merged to unblock next tasks, the api result should not change if we switch one method for the other later.

Reply:
yep. Let's merge it. And change it later if we want. At least you are aware of the alternative now :)
It's also possible that using oauth token would make simpler to enable not-enabled repos for example. Just keep it in mind.